### PR TITLE
update gas fees

### PR DESCRIPTION
**Pull request by segun**
Fixes: #13496

Explanation:  This PR uses a direct method to update transaction gas fees instead of calling the updateTransaction method.

Manual testing steps:  
  - 
  - 
  -

### CODE REVIEW

File: ui/hooks/gasFeeInput/useTransactionFunctions.js [hunk 1]
```diff
@@ -87,7 +88,11 @@ export const useTransactionFunctions = ({
         );
         dispatch(updateCustomSwapsEIP1559GasParams(newGasSettings));
       } else {
-        dispatch(updateTransactionFn(updatedTxMeta));
+        const userSettings = {};
+        userSettings.userEditedGasLimit = updatedTxMeta.userEditedGasLimit;
+        userSettings.userFeeLevel = updatedTxMeta.userFeeLevel;
+        dispatch(updateTransactionGasFees(updatedTxMeta.id, newGasSettings));
+        dispatch(updateTransactionUserSettings(updatedTxMeta.id, userSettings));
```
Comment (jpuri) [inline ui/hooks/gasFeeInput/useTransactionFunctions.js:95]: `txMeta` may have more fields, do we ensure that all is saved ?

https://github.com/MetaMask/metamask-extension/pull/13646/files#diff-a37b7544a0b876163dca658b96d70b11e99c7d7c33cb64ef2295c2244343b981R73

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -568,7 +568,6 @@ export default class TransactionController extends EventEmitter {
     if (!this._checkIfTxStatusIsUnapproved(txId)) {
       return;
     }
-
```
Comment (darkwing) [inline app/scripts/controllers/transactions/index.js:571]: This change is unrelated and can probably be reverted.

**Review by darkwing — CHANGES_REQUESTED** [review]

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 1 above)
Comment (segun) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:95]: @jpuri The essence of this change is that we don't want to update `all` fields, just the ones that should be updated. In this case the gas settings.

Comment (segun) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:95]: @darkwing can't do that because of [this](https://eslint.org/docs/rules/no-case-declarations)

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 3]
```diff
@@ -187,7 +189,11 @@ export default function EditGasPopover({
         );
         break;
       case EDIT_GAS_MODES.MODIFY_IN_PLACE:
-        dispatch(updateTransaction(updatedTxMeta));
+        userSettings.userEditedGasLimit = updatedTxMeta.userEditedGasLimit;
+        userSettings.userFeeLevel = updatedTxMeta.userFeeLevel;
+
+        dispatch(updateTransactionGasFees(updatedTxMeta.id, newGasSettings));
+        dispatch(updateTransactionUserSettings(updatedTxMeta.id, userSettings));
```
[the remark this answers is not in the document]
Comment (segun) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:196]: @darkwing can't do that because of [this](https://eslint.org/docs/rules/no-case-declarations)

File: app/scripts/controllers/transactions/index.js (hunk 2 above)
Comment (segun) [reply app/scripts/controllers/transactions/index.js:571]: fixed here [4f107a9](https://github.com/MetaMask/metamask-extension/pull/13646/commits/4f107a903a983d58be8f2b905927939894804c4e)

File: ui/ducks/metamask/metamask.js [hunk 4]
```diff
@@ -209,7 +209,7 @@ const toHexWei = (value, expectHexWei) => {
 };
 
 // Action Creators
-export function updateTransactionGasFees({
+export function updateTransaction({
```
Comment (brad-decker) [inline ui/ducks/metamask/metamask.js:212]: [Q] Why is this function being renamed? I am not a fan of this change because it mimics the naming of a function that previously updated the transaction in state in the back-end... which made me think all of its usages in this were that old function.  And it seems like this is still only updating gas related fields? So why was the name changed?

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/ducks/metamask/metamask.js (hunk 4 above)
Comment (segun) [reply ui/ducks/metamask/metamask.js:212]: I renamed it to avoid name collision (in this file) with updateTransactionGasFees of txController.

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 1 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:95]: @segun : but in this case there are more fields than gas settings in txMeta which will be lost.

**Review by jpuri — CHANGES_REQUESTED** [review]
This change can effect cancel / speed-up flow: https://github.com/MetaMask/metamask-extension/pull/13646/files#r814856524

File: ui/ducks/metamask/metamask.js (hunk 4 above)
Comment (segun) [reply ui/ducks/metamask/metamask.js]: I changed the name to `updateGasFees` [here](https://github.com/MetaMask/metamask-extension/pull/13646/commits/70f994ad10acfb26327d4dd1398c0195e91ab258)

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 1 above)
Comment (segun) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:95]: @jpuri I am updating gas fees and userSettingsFees based on 

1. the parameters passed to `useTransactionFunction`
2. the computed values in the method.

File: ui/hooks/gasFeeInput/useTransactionFunctions.js [hunk 5]
```diff
@@ -87,7 +89,18 @@ export const useTransactionFunctions = ({
...
+        userSettings.userEditedGasLimit = updatedTxMeta.userEditedGasLimit;
+        userSettings.userFeeLevel = updatedTxMeta.userFeeLevel;
+
+        if (txMeta && txMeta.previousGas) {
+          dispatch(
+            updatePreviousGasParams(updatedTxMeta.id, txMeta.previousGas),
+          );
+        }
+
+        dispatch(updateTransactionGasFees(updatedTxMeta.id, newGasSettings));
```
Comment (danjm) [inline ui/hooks/gasFeeInput/useTransactionFunctions.js:102]: I think we need to `await` the first dispatch before we call the second

File: ui/hooks/gasFeeInput/useTransactionFunctions.js [hunk 6]
```diff
@@ -87,7 +89,18 @@ export const useTransactionFunctions = ({
         );
         dispatch(updateCustomSwapsEIP1559GasParams(newGasSettings));
       } else {
-        dispatch(updateTransactionFn(updatedTxMeta));
+        const userSettings = {};
+        userSettings.userEditedGasLimit = updatedTxMeta.userEditedGasLimit;
+        userSettings.userFeeLevel = updatedTxMeta.userFeeLevel;
+
+        if (txMeta && txMeta.previousGas) {
+          dispatch(
```
Comment (danjm) [inline ui/hooks/gasFeeInput/useTransactionFunctions.js:97]: We should await this

File: ui/store/actions.js [hunk 7]
```diff
@@ -673,6 +673,70 @@ const updateMetamaskStateFromBackground = () => {
   });
 };
 
+export function updateTransactionGasFees(txId, txGasFees) {
+  return async (dispatch) => {
+    try {
+      await promisifiedBackground.updateTransactionGasFees(txId, txGasFees);
+    } catch (error) {
+      dispatch(txError(error));
+      dispatch(goHome());
```
Comment (danjm) [inline ui/store/actions.js:682]: I don't think we need `dispatch(goHome());` anymore.

File: ui/store/actions.js [hunk 8]
```diff
@@ -673,6 +673,70 @@ const updateMetamaskStateFromBackground = () => {
...
+      log.error(error.message);
+      throw error;
+    }
+
+    dispatch(updateTransactionParams(txGasFees.id, txGasFees.txParams));
+    const newState = await updateMetamaskStateFromBackground();
+    dispatch(updateMetamaskState(newState));
+    dispatch(showConfTxPage({ id: txId }));
```
Comment (danjm) [inline ui/store/actions.js:690]: We don't need `dispatch(showConfTxPage({ id: txId }));` anymore

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 1 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:103]: @danjm : in EIP-1559 V2 we added this field `previousGas` in txMeta. This is used in case of cancel/speedup and saves the gas used in original transaction. This field is used to derive the new estimate type `10% Minimum` for the transaction.

Commit here [fixes](https://github.com/MetaMask/metamask-extension/pull/13646/commits/83652f84fa8a631e6bcabf8a0af2d4ef9d5159a9#diff-675c437262d906486cf5436c0b64535347286cc4d1081dbd863fb45b8a01b81aR386) this issue.

File: ui/store/actions.js (hunk 8 above)
Comment (danjm) [reply ui/store/actions.js:690]: same for the other methods as well

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 6 above)
Comment (segun) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js]: Fixed [here](https://github.com/MetaMask/metamask-extension/pull/13646/commits/1669aac0914ec5681b651a1e3a76546f5108bd3e)

File: ui/hooks/gasFeeInput/useTransactionFunctions.js (hunk 5 above)
Comment (segun) [reply ui/hooks/gasFeeInput/useTransactionFunctions.js:102]: Fixed [here](https://github.com/MetaMask/metamask-extension/pull/13646/commits/1669aac0914ec5681b651a1e3a76546f5108bd3e)

File: app/scripts/controllers/transactions/index.js [hunk 9]
```diff
@@ -406,6 +436,17 @@ export default class TransactionController extends EventEmitter {
...
+   * =======
+   * {
+   * gasLimit,
+   * gasPrice,
+   * maxPriorityFeePerGas,
+   * maxFeePerGas,
+   * estimateUsed,
+   * estimateSuggested
+   * }
+   * >>>>>>> 83652f84f (fix:)
```
Comment (jpuri) [inline app/scripts/controllers/transactions/index.js:449]: 🔪

File: test/e2e/metamask-ui.spec.js [hunk 10]
```diff
@@ -279,6 +279,7 @@ describe('MetaMask', function () {
       await gasPriceInput.fill('20');
       await driver.delay(veryLargeDelayMs);
       await driver.clickElement({ text: 'Save', tag: 'button' });
+      await driver.delay(veryLargeDelayMs);
```
Comment (segun) [inline test/e2e/metamask-ui.spec.js:282]: You'll see some of these pattern in e2e tests in this PR where I have to call delay after Save is clicked. This is because it takes some milliseconds between when Save is clicked and when the values are updated and if the next line is executed immediately, it will use the old values.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 11]
```diff
@@ -187,7 +189,13 @@ export default function EditGasPopover({
         );
         break;
```
Comment (jpuri) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:190]: Hey @segun : sorry for this delayed feedback, but the current update transaction function results in loading spinner being shown on UI till changes are reflected in app state. Will there be same behaviour with these new update methods.

`userFeeLevel` and `gas values` are kind of related, I think it would be better to update them in single method. `await` below is helping in achieving this. But if we can have a single methods at controller level, that will be even  better.

Comment (segun) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:190]: There was a comment (a long time ago...I will try and find it) that asked that I remove the spinners.

On 2. Yeah, I can have a method that combine both. But again the crux of the PR is to separate these calls. 

@danjm what do you think about combining these 2?

Comment (danjm) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:190]: @segun I think combining them is fine. What's most important is the removal of the `updateTransaction` function, which has the ability to update _every_ parameter. But I think it is fine to combine these

Comment (jpuri) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:190]: Also, for edit gas fee popup, spinner is important, as this spinner shows up on confirm transaction page. It is important for this spinner to be visible till estimates are updated.

Comment (danjm) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:190]: okay, @segun let's re-add the logic for showing the spinner

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 12]
```diff
@@ -187,7 +189,14 @@ export default function EditGasPopover({
...
       case EDIT_GAS_MODES.MODIFY_IN_PLACE:
-        dispatch(updateTransaction(updatedTxMeta));
+        newGasSettings.userEditedGasLimit = updatedTxMeta.userEditedGasLimit;
+        newGasSettings.userFeeLevel = updatedTxMeta.userFeeLevel;
+
+        dispatch(showLoadingIndication());
+        await dispatch(
+          updateTransactionGasFees(updatedTxMeta.id, newGasSettings),
+        );
+        dispatch(hideLoadingIndication());
```
Comment (jpuri) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:199]: It will be nice to wrap this whole thing 
```
dispatch(showLoadingIndication());
        await dispatch(
          updateTransactionGasFees(updatedTxMeta.id, newGasSettings),
        );
        dispatch(hideLoadingIndication());
```
in hook so that all place in UI which use this function will not have to replicate the logic to show / hide loader.

**Review by jpuri — DISMISSED** [review]

**Review by danjm — DISMISSED** [review]

**Review by darkwing — DISMISSED** [review]

**Review by brad-decker — DISMISSED** [review]

**Review by brad-decker — DISMISSED** [review]

**Review by brad-decker — DISMISSED** [review]

File: app/scripts/controllers/transactions/index.js [hunk 13]
```diff
@@ -378,14 +378,42 @@ export default class TransactionController extends EventEmitter {
    * updates the params that are editible in the send edit flow
    *
    * @param {string} txId - transaction id
-   * @param {object} editableParams - holds the editable parameters
+   * @param {object} previousGasParams - holds the parameter to update
+   * @param {string} previousGasParams.maxFeePerGas
+   * @param {string} previousGasParams.maxPriorityFeePerGas
+   * @param {string} previousGasParams.gasLimit
+   */
+  updatePreviousGasParams(
```
Comment (danjm) [inline app/scripts/controllers/transactions/index.js:386]: This method should be updated to reflect the changes I made in the other methods: returning the full updated txMeta

File: ui/store/actions.js [hunk 14]
```diff
@@ -671,6 +671,23 @@ const updateMetamaskStateFromBackground = () => {
   });
 };
 
+export function updatePreviousGasParams(txId, previousGasParams) {
```
Comment (danjm) [inline ui/store/actions.js:674]: This should be updated to reflect the changes I made in the other method, and return the updated transaction

**Review by danjm — COMMENTED** [review]
LGTM

**Review by darkwing — APPROVED** [review]
Make sense to me.  Might be good to get @PeterYinusa/QA's thoughts on the `driver.delay` calls.  Nice job @segun !

**Review by brad-decker — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

File: test/e2e/metamask-ui.spec.js (hunk 10 above)
Comment (PeterYinusa) [reply test/e2e/metamask-ui.spec.js:281]: In future cases, where possible we can use the following method to wait for an expected condition (e.g. updated gas estimate), instead of adding a delay

```
await driver.waitForSelector({
	css: 'GAS_ESTIMATE_CSS_SELECTOR',
	text: 'GAS_ESTIMATE_UPDATED_VALUE',
});
```
